Index: ffi/Cargo.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/ffi/Cargo.toml b/ffi/Cargo.toml
--- a/ffi/Cargo.toml	(revision 2334ab150abd803a555b1541a7e44891b7f5cc60)
+++ b/ffi/Cargo.toml	(date 1639035151918)
@@ -21,6 +21,7 @@
 libc = "0.2.42"
 num_enum = "0.5"
 serde_json = "1.0.56"
+percent-encoding = "2.1.0"

 [dependencies.wirefilter-engine]
 path = "../engine"
Index: ffi/src/lib.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/ffi/src/lib.rs b/ffi/src/lib.rs
--- a/ffi/src/lib.rs	(revision 2334ab150abd803a555b1541a7e44891b7f5cc60)
+++ b/ffi/src/lib.rs	(date 1639042359703)
@@ -16,10 +16,7 @@
     io::{self, Write},
     net::IpAddr,
 };
-use wirefilter::{
-    AlwaysList, Array, ExecutionContext, FieldIndex, Filter, FilterAst, LhsValue, ListDefinition,
-    Map, NeverList, ParseError, Scheme, Type,
-};
+use wirefilter::{AlwaysList, Array, ExecutionContext, FieldIndex, Filter, FilterAst, LhsValue, ListDefinition, Map, NeverList, ParseError, Scheme, Type, DefaultCompiler, FunctionArgs, SimpleFunctionDefinition, SimpleFunctionParam, FunctionArgKind, SimpleFunctionImpl, List, AlwaysListMatcher, NeverListMatcher};

 const VERSION: &str = env!("CARGO_PKG_VERSION");

@@ -131,6 +128,169 @@

 type HashingResult = CResult<u64>;

+use wirefilter::types::GetType;
+use crate::transfer_types::raw_ptr_repr::ExternPtrRepr;
+
+fn len_function<'a>(args: FunctionArgs<'_, 'a>) -> Option<LhsValue<'a>> {
+    match args.next()? {
+        Ok(LhsValue::Bytes(bytes)) => Some(LhsValue::Int(i32::try_from(bytes.len()).unwrap())),
+        Err(Type::Bytes) => None,
+        _ => unreachable!(),
+    }
+}
+
+fn any_function<'a>(args: FunctionArgs<'_, 'a>) -> Option<LhsValue<'a>> {
+    match args.next()? {
+        Ok(v) => Some(LhsValue::Bool(
+            Array::try_from(v)
+                .unwrap()
+                .into_iter()
+                .any(|lhs| bool::try_from(lhs).unwrap()),
+        )),
+        Err(Type::Array(ref arr)) if arr.get_type() == Type::Bool => None,
+        _ => unreachable!(),
+    }
+}
+
+fn all_function<'a>(args: FunctionArgs<'_, 'a>) -> Option<LhsValue<'a>> {
+    match args.next()? {
+        Ok(v) => Some(LhsValue::Bool(
+            Array::try_from(v)
+                .unwrap()
+                .into_iter()
+                .all(|lhs| bool::try_from(lhs).unwrap()),
+        )),
+        Err(Type::Array(ref arr)) if arr.get_type() == Type::Bool => None,
+        _ => unreachable!(),
+    }
+}
+
+fn lower_function<'a>(args: FunctionArgs<'_, 'a>) -> Option<LhsValue<'a>> {
+    let input = args.next()?.ok()?;
+    match input {
+        LhsValue::Bytes(bytes) => Some(LhsValue::Bytes(bytes.to_ascii_lowercase().into())),
+        _ => panic!("Invalid type: expected Bytes, got {:?}", input),
+    }
+}
+
+fn upper_function<'a>(args: FunctionArgs<'_, 'a>) -> Option<LhsValue<'a>> {
+    let input = args.next()?.ok()?;
+    match input {
+        LhsValue::Bytes(bytes) => Some(LhsValue::Bytes(bytes.to_ascii_uppercase().into())),
+        _ => panic!("Invalid type: expected Bytes, got {:?}", input),
+    }
+}
+
+use percent_encoding::percent_decode;
+
+fn url_decode<'a>(args: FunctionArgs<'_, 'a>) -> Option<LhsValue<'a>> {
+    let input = args.next()?.ok()?;
+    let res = match input {
+        LhsValue::Bytes(bytes) => Some(LhsValue::Bytes(percent_decode(&bytes).collect())),
+        _ => panic!("Invalid type: expected Bytes, got {:?}", input),
+    };
+    return res;
+}
+
+#[no_mangle]
+pub extern "C" fn add_standard_functions(
+    scheme: &mut Scheme,
+) {
+    scheme
+        .add_function(
+            "any".into(),
+            SimpleFunctionDefinition {
+                params: vec![SimpleFunctionParam {
+                    arg_kind: FunctionArgKind::Field,
+                    val_type: Type::Array(Box::new(Type::Bool)),
+                }],
+                opt_params: vec![],
+                return_type: Type::Bool,
+                implementation: SimpleFunctionImpl::new(any_function),
+            },
+        )
+        .unwrap();
+    scheme
+        .add_function(
+            "all".into(),
+            SimpleFunctionDefinition {
+                params: vec![SimpleFunctionParam {
+                    arg_kind: FunctionArgKind::Field,
+                    val_type: Type::Array(Box::new(Type::Bool)),
+                }],
+                opt_params: vec![],
+                return_type: Type::Bool,
+                implementation: SimpleFunctionImpl::new(all_function),
+            },
+        )
+        .unwrap();
+    scheme
+        .add_function(
+            "lower".into(),
+            SimpleFunctionDefinition {
+                params: vec![SimpleFunctionParam {
+                    arg_kind: FunctionArgKind::Field,
+                    val_type: Type::Bytes,
+                }],
+                opt_params: vec![],
+                return_type: Type::Bytes,
+                implementation: SimpleFunctionImpl::new(lower_function),
+            },
+        )
+        .unwrap();
+    scheme
+        .add_function(
+            "upper".into(),
+            SimpleFunctionDefinition {
+                params: vec![SimpleFunctionParam {
+                    arg_kind: FunctionArgKind::Field,
+                    val_type: Type::Bytes,
+                }],
+                opt_params: vec![],
+                return_type: Type::Bytes,
+                implementation: SimpleFunctionImpl::new(upper_function),
+            },
+        )
+        .unwrap();
+    scheme
+        .add_function(
+            "len".into(),
+            SimpleFunctionDefinition {
+                params: vec![SimpleFunctionParam {
+                    arg_kind: FunctionArgKind::Field,
+                    val_type: Type::Bytes,
+                }],
+                opt_params: vec![],
+                return_type: Type::Int,
+                implementation: SimpleFunctionImpl::new(len_function),
+            },
+        )
+        .unwrap();
+    scheme
+        .add_function(
+            "url_decode".into(),
+            SimpleFunctionDefinition {
+                params: vec![SimpleFunctionParam {
+                    arg_kind: FunctionArgKind::Field,
+                    val_type: Type::Bytes,
+                }],
+                opt_params: vec![],
+                return_type: Type::Bytes,
+                implementation: SimpleFunctionImpl::new(url_decode),
+            },
+        )
+        .unwrap();
+}
+
+#[no_mangle]
+pub extern "C" fn set_nevermatch_iplist(
+    exec_context: &mut ExecutionContext<'_>,
+) -> bool {
+    let list = exec_context.scheme().get_list(&Type::Ip).unwrap();
+    exec_context.set_list_matcher(list, NeverListMatcher {}).unwrap();
+    return true;
+}
+
 #[no_mangle]
 pub extern "C" fn wirefilter_create_scheme() -> RustBox<Scheme> {
     Default::default()
Index: engine/src/lib.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/engine/src/lib.rs b/engine/src/lib.rs
--- a/engine/src/lib.rs	(revision 2334ab150abd803a555b1541a7e44891b7f5cc60)
+++ b/engine/src/lib.rs	(date 1639034868536)
@@ -73,7 +73,7 @@
 mod rhs_types;
 mod searcher;
 mod strict_partial_ord;
-mod types;
+pub mod types;

 pub use self::{
     ast::{
Index: engine/src/list_matcher.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/engine/src/list_matcher.rs b/engine/src/list_matcher.rs
--- a/engine/src/list_matcher.rs	(revision 2334ab150abd803a555b1541a7e44891b7f5cc60)
+++ b/engine/src/list_matcher.rs	(date 1639042607772)
@@ -174,7 +174,7 @@
     }

     fn is_valid_matcher(&self, matcher: &dyn Any) -> bool {
-        matcher.is::<AlwaysListMatcher>()
+        matcher.is::<NeverListMatcher>()
     }
 }

